### DISCUSSION TITLE
Using valhalla as a routing engine in offline c++ project

### DISCUSSION THREAD

**Discussion by abidbodal** [valhalla/valhalla#2887] [Q&A]
I would like to use the valhalla routing engine in a C++ based mapping application to create offline realtime routing. 

I looked at the demo examples and have most of them working, but they are mostly setup to make an HTTP request to the valhalla_service.

Is there any example that uses the valhalla engine as a c++ library to do realtime routing? I built and installed valhalla on a mac but now when I link to it (by including headers), I am seeing thirdparty headers such as date.h missing in /usr/local/include/valhalla. 

Currently I am testing on my mac, but it will eventually be put on a Linux Yocto based ARM7 processor

Any advice on using Valhalla as a c++ library would be really helpful.

**Comment by kevinkreiser** [comment]
@abidbodal if we are referencing includes that we dont then put in our installed package that is a problem. basically ever since we switched to cmake we dropped proper packaging support. we have install targets but because we dont package binaries for PPA we dont force ourselves to make sure that the required installation targets cover everything thats referenced. sadly, over time we've let some headers slip through the cracks.

so first, as you rightly called out, we should figure out what missing headers are referenced in our headers and make sure they are marked as PUBLIC targets in our cmake files. honestly though it would be best if we can pimpl as many of those as possible, its a much better thing when they become implementation details rather than API interface. if you like you can list the headers here and we can PR fixes to that and reference this issue.

after that, it seems like you are already on the right track to using it as a library. basically it should install the libs and the includes and  you should be able to start linking libvalhalla in your application. we also support python bindings so if you want to do some scripting that can work as well. another interesting way of running valhalla is to actually build the server and run the server on the "offline" device. this is useful when you have a device that comes on and offline depending on where it is. when its online it hits one url and when its offline it hits localhost, makes the "clientside" code simpler to always be using http and just changing the base of the url.

**Reply by abidbodal** [reply]
@kevinkreiser I appreciate the quick response.

Our device is totally offline - bicycle computer. The only connection it has to the `outside world` is BT and USB, and Bluetooth is not fast enough to transfer much content and USB is used to load tiles on initial setup of the device.

I was avoiding python binding because the device currently doesn't have python support, but that could be an option.

I was thinking of creating a service that receives protobuf events and responds back with protobuf events. Is there a good example of this? I am trying to follow valhalla_service.cc as an example and skipping the json parsing part.

My current build errors are due to third_party headers in `thirdparty/date`  and `thirdparty/rapidjson`

**Comment by kevinkreiser** [comment]
@abidbodal yes, i would suggest using the `actor` interface. its a single header that you'll need to include to use it to get what you need going. at the moment it takes json strings as the requests and returns json strings as the output but it does allow you to also pass in a protobuf object so that you can also get the result as pbf too. there is currently no interface for providing protobuf as the input though i'll be working on allowing that shortly i reckon.

for an example of how to do something like this in code you can see the one-shot mode we have in the service: https://github.com/valhalla/valhalla/blob/master/src/valhalla_service.cc#L51-L138

if you do something like that you'll be pretty much good to go. if you want to get back protobuf you need to make a `valhalla::Api` object and pass a pointer to that to the respective actor function that is called. another trick is you'll want ot keep the actor object alive for as long as you can because instantiating a new one (with a new graphreader at least) is expensive (clears cache and has to scan files or directories depending on how you have it configured).

**Reply by abidbodal** [reply]
@kevinkreiser I have been trying to get `valhalla_service.cc` to build as a separate project by linking to `libvalhalla.a` and including all the header files in valhalla/valhalla/ but I keep getting the following error. Can you point me to how libvalhalla.a is being generated? Is it being generated for a different architecture? I keep getting errors similar to this:

```
/Library/Developer/CommandLineTools/usr/bin/c++  -g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/routing_lib_example.dir/main.cpp.o  -o routing_lib_example  ../lib/libvalhalla.a 
Undefined symbols for architecture x86_64:
  "prime_server::http_request_t::from_string(char const*, unsigned long)", referenced from:
      valhalla::loki::loki_worker_t::work(std::__1::list<zmq::message_t, std::__1::allocator<zmq::message_t> > const&, void*, std::__1::function<void ()> const&) in libvalhalla.a(worker.cc.o)
...
```

Here's a simple project I created which is essentially just a copy of valhalla_service.cc, includes and libvalhalla.a
https://github.com/abidbodal/routing_lib_example

I add libvalhalla using these lines in cmakelists.txt
```
include_directories(${Boost_INCLUDE_DIR}
        ${RAPIDJSON_INCLUDE_DIR}
        ${DATE_INCLUDE_DIR}
        ${CMAKE_SOURCE_DIR}/includes/valhalla
        )
add_dependencies(routing_lib_example rapidjson date)
target_link_libraries(routing_lib_example ${CMAKE_SOURCE_DIR}/lib/libvalhalla.a)```

**Comment by nilsnolde** [comment]
@abidbodal 

when you get the ARM build to work, it would be cool if you could share the recipe. We've recently deprecated the only 32-bit CI build (was a Win32 build) and are eyeing an ARM CI build as replacement. Or even a PR from you for a Github Action would be super great of course:) Could also be your insurance that Valhalla isn't doing anything which screws with 32bit support.

**Comment by kevinkreiser** [comment]
@abidbodal yeah the main issue there is that the service bits of the library depend on another library called `prime_server`. it seems to me that your static valhalla library is still linking the dynamic prime_server library. we should be able to link prime_server statcally via some changes to cmake but that said there arent any ppa builds for ARM of prime_server. honestly though if i were you, and if i were working on the arm build i would be disabling the http services with `-DENABLE_SERVICES=Off` as a cmake option so that you dont depend on prime_server at all.

**Reply by abidbodal** [reply]
@kevinkreiser Apologize for the super late response. I was working on a different project for a bit.

I set the following options on the build and then copied libvalhalla.a into the project and got the following linker error:
```
option(ENABLE_TOOLS "Enable Valhalla tools" OFF)
option(ENABLE_DATA_TOOLS "Enable Valhalla data tools" OFF)
option(ENABLE_SERVICES "Enable Valhalla services" OFF)
option(ENABLE_HTTP "Enable the use of CURL" OFF)
option(ENABLE_PYTHON_BINDINGS "Enable Python bindings" OFF)
option(ENABLE_TESTS "Enable Valhalla tests" OFF)
option(ENABLE_BENCHMARKS "Enable microbenchmarking" OFF)
```

Error
```"/Users/abidesmail/Library/Application Support/JetBrains/Toolbox/apps/CLion/ch-1/203.5981.166/CLion.app/Contents/bin/cmake/mac/bin/cmake" -E cmake_link_script CMakeFiles/routing_lib_example.dir/link.txt --verbose=1
/Library/Developer/CommandLineTools/usr/bin/c++  -g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/routing_lib_example.dir/main.cpp.o  -o routing_lib_example  ../lib/libvalhalla.a 
Undefined symbols for architecture x86_64:
  "google::protobuf::MessageLite::ParseFromCodedStream(google::protobuf::io::CodedInputStream*)", referenced from:
      (anonymous namespace)::incident_singleton_t::read_tile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libvalhalla.a(graphreader.cc.o)
  "google::protobuf::StringPiece::LogFatalSizeTooBig(unsigned long, char const*)", referenced from:
      valhalla::DirectionsLeg_GuidanceView_Type_Parse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, valhalla::DirectionsLeg_GuidanceView_Type*) in libvalhalla.a(directions.pb.cc.o)
      valhalla::DirectionsLeg_Maneuver_CardinalDirection_Parse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, valhalla::DirectionsLeg_Maneuver_CardinalDirection*) in libvalhalla.a(directions.pb.cc.o)
      valhalla::DirectionsLeg_Maneuver_Type_Parse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, valhalla::DirectionsLeg_Maneuver_Type*) in libvalhalla.a(directions.pb.cc.o)
      valhalla::DirectionsLeg_Maneuver_BssManeuverType_Parse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, valhalla::DirectionsLeg_Maneuver_BssManeuverType*) in libvalhalla.a(directions.pb.cc.o)
      valhalla::DirectionsLeg_TravelMode_Parse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, valhalla::DirectionsLeg_TravelMode*) in libvalhalla.a(directions.pb.cc.o)
      valhalla::DirectionsLeg_VehicleType_Parse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, valhalla::DirectionsLeg_VehicleType*) in libvalhalla.a(directions.pb.cc.o)
      valhalla::DirectionsLeg_PedestrianType_Parse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, valhalla::DirectionsLeg_PedestrianType*) in libvalhalla.a(directions.pb.cc.o)
      ...
  "google::protobuf::RepeatedField<unsigned int>::Reserve(int)", referenced from:
      valhalla::TripLeg_ShapeAttributes::TripLeg_ShapeAttributes(valhalla::TripLeg_ShapeAttributes const&) in libvalhalla.a(trip.pb.cc.o)
      valhalla::TripLeg_ShapeAttributes::_InternalParse(char const*, google::protobuf::internal::ParseContext*) in libvalhalla.a(trip.pb.cc.o)
      valhalla::TripLeg_ShapeAttributes::MergeFrom(valhalla::TripLeg_ShapeAttributes const&) in libvalhalla.a(trip.pb.cc.o)
      valhalla::IncidentsTile_Metadata::IncidentsTile_Metadata(valhalla::IncidentsTile_Metadata const&) in libvalhalla.a(incidents.pb.cc.o)
      valhalla::IncidentsTile_Metadata::_InternalParse(char const*, google::protobuf::internal::ParseContext*) in libvalhalla.a(incidents.pb.cc.o)
      valhalla::IncidentsTile_Metadata::MergeFrom(valhalla::IncidentsTile_Metadata const&) in libvalhalla.a(incidents.pb.cc.o)
      valhalla::thor::TripLegBuilder::Build(valhalla::Options const&, valhalla::thor::AttributesController const&, valhalla::baldr::GraphReader&, std::__1::array<std::__1::shared_ptr<valhalla::sif::DynamicCost>, 4ul> const&, std::__1::__wrap_iter<valhalla::thor::PathInfo const*>, std::__1::__wrap_iter<valhalla::thor::PathInfo const*>, valhalla::Location&, valhalla::Location&, std::__1::list<valhalla::Location, std::__1::allocator<valhalla::Location> > const&, valhalla::TripLeg&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::function<void ()> const*, std::__1::unordered_map<unsigned long, std::__1::pair<valhalla::thor::EdgeTrimmingInfo, valhalla::thor::EdgeTrimmingInfo>, std::__1::hash<unsigned long>, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::pair<valhalla::thor::EdgeTrimmingInfo, valhalla::thor::EdgeTrimmingInfo> > > >*) in libvalhalla.a(triplegbuilder.cc.o)
      ...```

Thanks

**Comment by purew** [comment]
```
Undefined symbols for architecture x86_64:
  "google::protobuf::MessageLite
```

It is telling you that it cant find the protobuf symbols that are defined in the library. You'll want to install the protobuf library or link protobuf statically (which I don't think we support currently).

**Comment by kevinkreiser** [comment]
@PureW it almost looks like hes got an older version or something. in some versions some of the functions moved from regular to the lite runtime.

**Comment by abidbodal** [comment]
@PureW @kevinkreiser I installed Valhalla using `sudo make install` (from the valhalla build directory) and then added the following to CMakeLists.txt
```include(FindProtobuf)
find_package(Protobuf REQUIRED)
include_directories(${PROTOBUF_INCLUDE_DIR})
target_link_libraries(routing_lib_example valhalla ${PROTOBUF_LIBRARY})```


Now I get the following error.
```"/Users/abidesmail/Library/Application Support/JetBrains/Toolbox/apps/CLion/ch-1/203.5981.166/CLion.app/Contents/bin/cmake/mac/bin/cmake" -E cmake_link_script CMakeFiles/routing_lib_example.dir/link.txt --verbose=1
/Library/Developer/CommandLineTools/usr/bin/c++  -g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/routing_lib_example.dir/main.cpp.o  -o routing_lib_example  /usr/local/lib/libvalhalla.a /usr/local/lib/libprotobuf.dylib /usr/local/lib/libvalhalla.a /usr/local/lib/libvalhalla.a 
Undefined symbols for architecture x86_64:
  "_deflate", referenced from:
      valhalla::baldr::deflate(std::__1::function<int (z_stream_s&)> const&, std::__1::function<void (z_stream_s&)> const&, int, bool) in libvalhalla.a(compression_utils.cc.o)
  "_deflateEnd", referenced from:
      valhalla::baldr::deflate(std::__1::function<int (z_stream_s&)> const&, std::__1::function<void (z_stream_s&)> const&, int, bool) in libvalhalla.a(compression_utils.cc.o)
  "_deflateInit2_", referenced from:
      valhalla::baldr::deflate(std::__1::function<int (z_stream_s&)> const&, std::__1::function<void (z_stream_s&)> const&, int, bool) in libvalhalla.a(compression_utils.cc.o)
  "_inflate", referenced from:
      valhalla::baldr::inflate(std::__1::function<void (z_stream_s&)> const&, std::__1::function<int (z_stream_s&)> const&) in libvalhalla.a(compression_utils.cc.o)
  "_inflateEnd", referenced from:
      valhalla::baldr::inflate(std::__1::function<void (z_stream_s&)> const&, std::__1::function<int (z_stream_s&)> const&) in libvalhalla.a(compression_utils.cc.o)
  "_inflateInit2_", referenced from:
      valhalla::baldr::inflate(std::__1::function<void (z_stream_s&)> const&, std::__1::function<int (z_stream_s&)> const&) in libvalhalla.a(compression_utils.cc.o)```

**Reply by kevinkreiser** [reply]
those are from zlib which the library also requires. perhaps it means that our CMakelists.txt doesnt properly set these as runtime dependencies for libvalhalla?

**Reply by abidbodal** [reply]
Thanks @kevinkreiser and @PureW . After adding ZLIB the application builds and runs. Now onto testing and using valhalla. Then I got to get all this to work on an ARM build

```
include("${CMAKE_SOURCE_DIR}/vendor/rapidjson.cmake")
include("${CMAKE_SOURCE_DIR}/vendor/date.cmake")
include_directories(${Boost_INCLUDE_DIR}
        ${RAPIDJSON_INCLUDE_DIR}
        ${DATE_INCLUDE_DIR}
        ${CMAKE_SOURCE_DIR}/includes/valhalla
        /usr/local/include
        )
add_dependencies(routing_lib_example rapidjson date)

include(FindProtobuf)
find_package(Protobuf REQUIRED)
find_package(ZLIB)
include_directories(${PROTOBUF_INCLUDE_DIR})
target_link_libraries(routing_lib_example
        ${CMAKE_SOURCE_DIR}/lib/libvalhalla.a
        ${PROTOBUF_LIBRARY}
        ZLIB::ZLIB)
```

**Reply by abidbodal** [reply]
I finally got back to the project and got the ARM32 build working. As I am testing, I am noticing that the test application I built (which is very similar to run_route) is using a lot of RAM - about 500MB. I sometimes get the following error:
`{"error_code":11080216,"error":":std::bad_alloc","status":"","status_code":11121123}`

which I think indicates I am running out of RAM. Is there a way I can avoid loading up as many tiles in RAM?

**Comment by kevinkreiser** [comment]
@abidbodal what size route are you computing and how are you loading the tiles, are you memory mapping them with a tar file or are you  using the tile_dir? you can set the limits of how much ram you use via a few different configurations but they are soft limits, if you have enough data and you try to plot a route that is long enough the router will try to route it.

we do have another type of tile cache that uses an LRU strategy and can be used to keep the memory requirements lower but we need some code changes in the algorithms to adhere to it. i need to do a bit of code review to make a decent suggestion to you after you explain a little bit more about whats going on in your use of the service.

**Reply by abidbodal** [reply]
The route I am computing is super short actually, here's the input to the run_route progam.
`./routing_lib_example ./valhalla_colorado_dash.json route '{"locations":[{"lat":39.864705,"lon":-105.224540,"type":"break"},{"lat":39.866896,"lon":-105.218554,"type":"break"}],"costing":"bicycle","costing_options":{"bicycle":{"bicycle_type":"road"}},"directions_options":{"units":"miles"}}'
`

I was loading up the tiles by providing the tar file. So maybe that's the problem. The application is very similar to what I had checked onto github, just built for ARM:
https://github.com/abidbodal/routing_lib_example

Eventually, this will not be the application we use, but I was just testing the arm library I built. The colorado tiles are 307.5MB.

**Comment by kevinkreiser** [comment]
how big is the tar? in tar mode we memory map the whole file, so if you dont have a lot of ram this will take a huge chunk of it. in ram constrained situations dont use the memory mapped tar. there is a lot of unwritten "wisdom" about configuration that it would be great for us to demystify at some point, apologies for the difficulties here!

**Reply by abidbodal** [reply]
@kevinkreiser How do I tell valhalla not to use the compressed file  and use a directory with uncompressed tiles, using the config.json file?

I tried setting "tile_dir": "/data/valhalla_tiles_colorado" and "tile_extract" : "", but that didn't work. Also tried removing "tile_extract" variable. Here's my current mjolnir settings 
 `"mjolnir": {                   
    "admin": "/data/valhalla_tiles_colorado/admins.sqlite",
    "data_processing": {                                   
      "allow_alt_name": false,                             
      "apply_country_overrides": true,                     
      "infer_internal_intersections": true,                
      "infer_turn_channels": true,                         
      "use_admin_db": true,                                
      "use_direction_on_ways": false,      
      "use_rest_area": false,              
      "use_urban_tag": false               
    },                                     
    "global_synchronized_cache": false,
    "hierarchy": true,                 
    "id_table_size": 1300000000,       
    "import_bike_share_stations": false,
    "include_bicycle": true,            
    "include_driveways": true,          
    "include_driving": true,            
    "include_pedestrian": true,         
    "logging": {                        
      "color": true,                    
      "file_name": "path_to_some_file.log",
      "type": "std_out"                    
    },                                     
    "lru_mem_cache_hard_control": false,   
    "max_cache_size": 1000000000,          
    "max_concurrent_reader_users": 1,      
    "reclassify_links": true,              
    "shortcuts": true,                  
    "tile_dir": "/data/valhalla_tiles_colorado",
    "tile_extract": "/data/valhalla_tiles_colorado.tar",
    "timezone": "/data/valhalla_tiles_colorado/timezones.sqlite",
    "traffic_extract": "/data/valhalla/traffic.tar",             
    "transit_dir": "/data/valhalla/transit",                     
    "use_lru_mem_cache": false,                                  
    "use_simple_mem_cache": false                                
  },`

**Comment by kevinkreiser** [comment]
Just don't create the tar in the first place. Or untar the tar into the directory where your config points. And a small correction, the tar is not compressed, it's just a way to pack all the tiles into one file rather than having many files on the disk

**Reply by abidbodal** [reply]
I set `tile_dir : /data/valhalla_tiles_colorado` and removed tile_extract from the config file. I also removed the tar file from /data/

`/data/valhalla_tiles_colorado# ls -al
drwxr-xr-x    5 root     root          8192 Jul 27 14:05 .
drwxr-xr-x   10 root     root          8192 Jul 27 15:30 ..
drwxr-xr-x    3 root     root          8192 Jul 27 14:05 0
drwxr-xr-x    5 root     root          8192 Jul 27 14:05 1
drwxr-xr-x    3 root     root          8192 Jul 27 14:05 2`

Now i'm getting the following error:2021/07/27 20:29:13.896270 [WARN] (stat): /data/valhalla/traffic.tar No such file or directory
`2021/07/27 20:29:13.898021 [WARN] Traffic tile extract could not be loaded
2021/07/27 20:29:13.985520 [WARN] /data/valhalla/elevation/ currently has no elevation tiles
2021/07/27 20:29:16.637414 [INFO] algorithm::bidirectional_a*
{"error_code":11272728,"error":":std::bad_alloc","status":"","status_code":11313635}`

**Reply by kevinkreiser** [reply]
ok well, at this point i think it would be good to confirm which allocation is happening. by default we reserve a bunch of ram for doing routing that doesnt have anything to do with the tile data but rather the structures needed to run the algorithm. it is possible to reserve less ram for those up front. it looks like by default we preallocate at least 80mb just for the edge labels. you can control that parameter with a config option. set the following config item in your config much lower: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config#L96

if that doesnt help then it might be another preallocation somewhere (since we do prealloc for all algorithms and there are about 6 different ones at this point) or it could be the tiles themselves. to find out we need to debug the code. let me know if you are up for that. the basic gist is:

- build it with `-DCMAKE_BUILD_TYPE=Debug` in the cmake step
- then prepend `gdb --args` to command you are using to start the server, that will take you to a gdb console
- in the gdb console type `catch throw`, which will tell gdb to stop the debugger on all exceptions
- then type `run` to start the program
- then issue your request and gdb should stop the debugger at the point when it throws the out of memory exception
- then you can simply type `where` and that will show us where its happening, copy paste that to this issue

**Reply by abidbodal** [reply]
It's strange, but I reset the device and remounted the /data/ partition and now I am NOT getting the allocation failures.

The colorado tarred tiles are about 300MB
Now with tarred tiles and `max_reserved_labels_count: 1000000`, VSZ usage is `467MB`
With untarred folder and `max_reserved_labels_count': 1000000`, VSZ usage is `197MB`

With tarred and untarred folder with `max_reserved_labels_count: 80000`, VSZ usage is `197MB`

So by setting `max_reserved_labels_count`, I can reduce the memory usage a bit. Our device has 250MB of physical memory, so we could be hitting the boundary pretty easily.

**Reply by abidbodal** [reply]
Thanks for the GDB commands. If I see the allocation failure again, i'll definitely setup GDB and send you logs.

**Reply by abidbodal** [reply]
@kevinkreiser What are Valhalla's memory requirements? I am using the actor interface and with colorado tiles loaded, I am seeing about 60MB of memory usage when a short route is being calculated

Are there any other strategies we can use to reduce the memory consumption? We currently have only 50MB of RAM available without a routing engine.

**Reply by kevinkreiser** [reply]
50 megabytes and some smaller portion of that that you can use to run the router with. whats the hardware platform, a toaster? jokes aside, i think to get to a place where that can happen we need to enumerate the consumers of memory. in my mind there are 2 such places to concentrate on:

1. the datastructures we use inside the algorithms to keep track of state when doing the path finding operation and when and how much we preallocate for those
2. the tile cache

Assuming the latter is the biggest user of ram (you can test this by starting the server or instantiating the "actor" class and see what ram usage is before running a route) I think the biggest thing we could do is add an option to trim the tile cache at the end of every iteration of the main loop of each algorithm (bidirectional a* for example). I think this should be pretty straight forward but there could be gotchas. I can make an example PR of it if you'd like to try it out and report back.

**Reply by abidbodal** [reply]
LOL - a bicycle computer. The current apps and OS use less than 150MB and the maps (which is Tangram based) uses 20MB (without tiles) and 35MB with Colorado Tiles.

Yeah, I would love to test out a PR. I think option 2 sounds good. In the bicycle routing use case, I would expect to be testing whether the user is off-route every couple of seconds and then possibly recomputing the route. Are there good strategies for testing when the cyclist is off-route?. We are currently thinking through what this algorithm should be.

I did a quick test on my mac and for the routingLibExample(based on valhalla service code without HTTP) uses 16MB when a route is not being computed and about 64MB when routing. Strangely we use a bit more on the ARM target. What's the best way to profile this? I just put a sleep right before the route generation and used `top -l 1|grep "routing"|awk '{print "MEM="$8 "\tRPRVT="$10}'` to measure the RAM usage - kinda a crude way to do this, maybe setting up valgrind might be better.

**Reply by kevinkreiser** [reply]
great! first thing i would do if i were you is disable building hierarchies when you make your tileset. the bicycle costing model doesnt use hierarchy culling so there is no sense to having more tile levels and more tiles in general. this will definitely reduce the size of your tile data by some amount :smile: but i've not tested it to see by how much.

second we want to use the LRU based tile cache. there are some settings in the config to turn this on and also set your limits in bytes. finally we should see if we can actually get away with calling trim in the main loop of the algorithm. id put something like `reader.trim()` round about: https://github.com/valhalla/valhalla/blob/master/src/thor/bidirectional_astar.cc#L512
